tinysearch/src/main.rs

```diff
@@ -0,0 +1,410 @@
+        let filters = Storage::from_bytes(&bytes)?.filters;
+        let results = base_search(&filters, self.term.clone(), self.num_searches);
+        for result in results {
+            println!(
+                "Title: {}, Url: {}, Meta: {:?}",
+                result.0, result.1, result.2
+            );
+        }
+        Ok(())
+    }
+}
+
+#[derive(Default)]
+struct Storage {
+    posts_index: PathBuf,
+    out_path: PathBuf,
+}
+
+impl Stage for Storage {
+    fn from_opt(opt: &Opt) -> Result<Self, Error> {
+        Ok(Self {
+            posts_index: opt.input_file.clone().context("No input file")?,
+            out_path: ensure_exists(opt.out_path.clone())?,
+        })
+    }
+
+    fn build(self: &Self) -> Result<(), Error> {
+        let storage_file = self.out_path.join("storage");
+        println!(
+            "Creating storage file for posts {} in file {}",
+            self.posts_index.display(),
+            storage_file.display()
+        );
+        let posts: Posts = index::read(fs::read_to_string(&self.posts_index)?)?;
+        trace!("Generating storage from posts: {:#?}", posts);
+        storage::write(posts, &storage_file)?;
+        println!("Storage ready in file {}", storage_file.display());
+        Ok(())
+    }
+}
+
+#[derive(Default)]
+struct Crate {
+    s: Storage,
+    out_path: PathBuf,
+    crate_name: String,
+    engine_version: toml_edit::Table,
+    non_top_level: bool
+}
+
+impl Stage for Crate {
+    fn from_opt(opt: &Opt) -> Result<Self, Error> {
+        let out_path = ensure_exists(opt.out_path.clone())?;
+        let storage_opt = {
+            let mut ret: Opt = opt.clone();
+            ret.out_path = ensure_exists(out_path.join("src"))?;
+            ret
+        };
+
+        Ok(Self {
+            s: Storage::from_opt(&storage_opt)?,
+            out_path: out_path,
+            crate_name: opt.crate_name.clone(),
+            engine_version: opt.engine_version.clone(),
+            non_top_level: opt.non_top_level_crate
+        })
+    }
+
+    fn build(self: &Self) -> Result<(), Error> {
+        println!(
+            "Creating tinysearch implementation crate {} in directory {}",
+            self.crate_name,
+            self.out_path.display()
+        );
+        let cargo_toml = self.out_path.join("Cargo.toml");
+        let mut cargo_toml_contents = templates::CRATE_CARGO_TOML.parse::<Document>()?;
+        cargo_toml_contents["package"]["name"] = value(self.crate_name.clone());
+        cargo_toml_contents["dependencies"]["tinysearch-engine"] =
+            toml_edit::Item::Table(self.engine_version.clone());
+        if self.non_top_level{
+            cargo_toml_contents.as_table_mut().remove("workspace");
+            cargo_toml_contents.as_table_mut().remove("profile");
+            cargo_toml_contents.as_table_mut().remove("lib");
+            cargo_toml_contents["lib"] = toml_edit::table();
+        }
+        fs::write(&cargo_toml, cargo_toml_contents.to_string())?;
```

Comment:
This is where the values in the `Cargo.toml` are set

Reply:
Makes perfect sense, thanks.